proto/types/types.proto

```diff
@@ -0,0 +1,173 @@
+
+  uint32 total = 1;
+  bytes  hash  = 2;
+}
+
+// BlockID
+message BlockID {
+  option (gogoproto.populate) = true;
+  option (gogoproto.equal)    = true;
+
+  bytes         hash         = 1;
+  PartSetHeader parts_header = 2 [(gogoproto.nullable) = false];
+}
+
+// --------------------------------
+
+// Header defines the structure of a Tendermint block header.
+message Header {
+  option (gogoproto.populate) = true;
+  option (gogoproto.equal)    = true;
+  // basic block info
+  Version                   version  = 1 [(gogoproto.nullable) = false];
+  string                    chain_id = 2;
+  int64                     height   = 3;
+  google.protobuf.Timestamp time     = 4 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
+
+  // prev block info
+  BlockID last_block_id = 5 [(gogoproto.nullable) = false];
+
+  // hashes of block data
+  bytes last_commit_hash = 6;  // commit from validators from the last block
+  bytes data_hash        = 7;  // transactions
+
+  // hashes from the app output from the prev block
+  bytes validators_hash      = 8;   // validators for the current block
+  bytes next_validators_hash = 9;   // validators for the next block
+  bytes consensus_hash       = 10;  // consensus params for current block
+  bytes app_hash             = 11;  // state after txs from the previous block
+  bytes last_results_hash    = 12;  // root hash of all results from the txs from the previous block
+
+  // consensus info
+  bytes evidence_hash    = 13;  // evidence included in the block
+  bytes proposer_address = 14;  // original proposer of the block
+}
+
+message Version {
+  option (gogoproto.populate) = true;
+  option (gogoproto.equal)    = true;
+
+  uint64 block = 1;
+  uint64 app   = 2;
+}
+
+// Data contains the set of transactions included in the block
+message Data {
+  // Txs that will be applied by state @ block.Height+1.
+  // NOTE: not all txs here are valid.  We're just agreeing on the order first.
+  // This means that block.AppHash does not include these txs.
+  repeated bytes txs = 1;
+  // Volatile
+  bytes hash = 2 [(gogoproto.customname) = "hash"];
+}
+
+// Vote represents a prevote, precommit, or commit vote from validators for
+// consensus.
+message Vote {
+  SignedMsgType             type      = 1;
+  int64                     height    = 2;
+  int32                     round     = 3;
+  BlockID                   block_id  = 4 [(gogoproto.nullable) = false];  // zero if vote is nil.
+  google.protobuf.Timestamp timestamp = 5
+      [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
+  bytes validator_address = 6;
+  int32 validator_index   = 7;
+  bytes signature         = 8;
+}
+
+// DuplicateVoteEvidence contains evidence a validator signed two conflicting
+// votes.
+message DuplicateVoteEvidence {
+  Vote vote_a = 1;
+  Vote vote_b = 2;
+}
+
+message Evidence {
+  // option (cosmos_proto.interface_type) = "*github.com/tendermint/tendermint/types.Evidence";
```

Comment:
We won't be able to use this tool as it causes cyclical imports. We will have to manually write switch statements for the respective modules/types